arch/arm/src/stm32f7/stm32_tickless.c

```diff
@@ -988,22 +988,31 @@ int up_timer_start(FAR const struct timespec *ts)
 }
 #endif
 
+#ifdef CONFIG_SCHED_TICKLESS_ALARM
 int up_alarm_start(FAR const struct timespec *ts)
 {
   uint64_t tm = ((uint64_t)ts->tv_sec * NSEC_PER_SEC + ts->tv_nsec) /
                 NSEC_PER_TICK;
-  uint64_t counter = ((uint64_t)g_tickless.overflow << 32) |
-                     STM32_TIM_GETCOUNTER(g_tickless.tch);
-
-  g_tickless.last_alrm = tm;
+  irqstate_t flags;
 
-  int32_t diff = tm / NSEC_PER_TICK + counter;
+  flags = enter_critical_section();
 
   STM32_TIM_SETCOMPARE(g_tickless.tch, CONFIG_STM32F7_TICKLESS_CHANNEL, tm);
 
   stm32_tickless_ackint(g_tickless.channel);
   stm32_tickless_enableint(CONFIG_STM32F7_TICKLESS_CHANNEL);
 
+  g_tickless.pending = true;
+
+  uint64_t counter = ((uint64_t)g_tickless.overflow << 32) |
+                     STM32_TIM_GETCOUNTER(g_tickless.tch);
+
+  if (counter > tm)
+    {
+      stm32_interval_handler();
```

Comment:
@gregory-nutt Any recommendations on how to call the stm32_interval_handler in the interrupt context instead of the callers context so that the expiration works as it would if called on timer expiration?

Reply:
> Any recommendations on how to call the stm32_interval_handler in the interrupt context instead of the callers context so that the expiration works as it would if called on timer expiration?

You can't really get into the interrupt context without being in an interrupt.  Entering a critical section is about as close as you can get.

The only side effect is per Issue #1138  If that is a concern you can also disable pre-emption.

Reply:
@gregory-nutt 

The problem is, this function eventually calls nxsig_timeout, like it should, but because the task is still currently running, the signal is not actually set, which means the task eventually waits and since the timeout already happened, things are locked.

I'm assuming it would not be safe to change line 128 of sig_timedwait.c to allow the signal in the TSTATE_TASK_RUNNING state.

Reply:
So you are saying that nxsig_timeout() is being called before up_block_task()?  This is the only place that nxsig_timeout is referenced:

    358               /* Start the watchdog */
    359
    360               wd_start(rtcb->waitdog, waitticks,
    361                        nxsig_timeout, 1, wdparm.pvarg);
    362
    363               /* Now wait for either the signal or the watchdog, but
    364                * first, make sure this is not the idle task,
    365                * descheduling that isn't going to end well.
    366                */
    367
    368               DEBUGASSERT(NULL != rtcb->flink);
    369               up_block_task(rtcb, TSTATE_WAIT_SIG);

The watchdog should not occur because we are in a critical section.  We will stay in the critical section until rtcb is blocked in TSTATE_WAIT_SIG state.  I guess I don't understand you problem description.

Reply:
@gregory-nutt 
The issue is that the wd_start happening line 360 is being called with a very small waitticks (6 = 6us).

By the time we get into wd_start, call up_alarm_start() we wind up trying to schedule a time that has already passed. Before my changes, this just resulted in the work queue being locked. I suppose it would wakeup after a full overflow cycle. 

So my change is trying to basically expire the timer before it even starts.

Reply:
> So my change is trying to basically expire the timer before it even starts.

The only way that I can see that happening is in sched/wdog/wd_start.c:

    232 #ifdef CONFIG_SCHED_TICKLESS
    233   /* Cancel the interval timer that drives the timing events.  This will
    234    * cause wd_timer to be called which update the delay value for the first
    235    * time at the head of the timer list (there is a possibility that it
    236    * could even remove it).
    237    */
    238
    239   nxsched_cancel_timer();
    240 #endif

Before we set up the new interval timer, we have to cancel the current one.

But this would not effect the wdog being started at line 360 where wd_start(0 is called; that wdog is not yet in the active wdog list.  That does not happen until after nxsched_cancel_timer(); is called.

The symptoms are different but this is very similar to the problem of issue #1138.  In that case, as I call, wd_cancel() was being called and causing an expected timer expiration for the same reason.  But the #1138 problem caused and unexpected context switch.

Reply:
It seems like the bottom line is that you are setting up a timer that is too short for the system to process properly using nxsig_timedwait().  There should be protections in place to handle that case, but there are not.

You might consider using a wdog directly and not call nxsig_timedwait();  Perhaps that would set up the timer more quickly.  But this is not a clean solution.
 
Or this could be a timer resolution issue.  What is the LSB resolution of the system timer?

Reply:
@gregory-nutt I think the problem here is a little different.

The issue here is that I WANT the timer to expire. The core issue that I'm trying to solve is that when up_alarm_start is called with a timestamp that has already passed, there is an assumption that the alarm is going to fire. But if the time has already passed by the time we set the compare register, I must force the expiration so that the work_queue can immediately process that event.

So I need a safe way of expiring the timer, from within, the up_alarm_start call, such that the work immediately just keeps moving. Am I making any sense?

Reply:
> So I need a safe way of expiring the timer, from within, the up_alarm_start call, such that the work immediately just keeps moving. Am I making any sense?

This would probably require some new design.

Reply:
Why not program the hardware to trigger the timer interrupt immediately(set the match value a little biggger than the count) in this case?
Actually, the new timer(e.g. ARM generic timer https://developer.arm.com/architectures/learn-the-architecture/generic-timer/the-processor-timers) design to work as OS tick normally change the interrupt trigger condition to: count >= match.

Reply:
Thanks for the input @xiaoxiang781216 

Yes, that is the only way I can think to do it is to set the compare register again, but bigger so that we can ensure we get the interrupt.

Where is this new timer used?

Reply:
All ARM Cortex-A and Cortex-R series switch to Generic Timer.

Reply:
An alternative for a 6uS delay would by up_udelay(6).  Normally, up_udelay() is not recommended because it hogs the CPU.  But in this case, the delay is so short that the task calling sig_timedwait() cannot even suspend.  So, in this case, there is no particular hogging and up_udelay() should be a viable alternative.

Reply:
How's about set the minimum time it takes to write the comp reg + 1 tick (MIN_T). Then an t < MIN_T gets set to MIN_T. This is sure you will get the int.

Reply:
@davids5 

I think I've already got something that works. Sounds somewhat similar to what you proposed. Check out the latest changes and let me know what you think.

Reply:
@antmerlino Could this result in 16 or 32 bits (based on timer type) counts of the timer with interrupts off?

Reply:
```
358               /* Start the watchdog */
359
360               wd_start(rtcb->waitdog, waitticks,
361                        nxsig_timeout, 1, wdparm.pvarg);
362
363               /* Now wait for either the signal or the watchdog, but
364                * first, make sure this is not the idle task,
365                * descheduling that isn't going to end well.
366                */
367
368               DEBUGASSERT(NULL != rtcb->flink);
369               up_block_task(rtcb, TSTATE_WAIT_SIG);
```

So my changes fix the issue at the tickless driver level, however now there is another race condition in sig_timedwait.c.

The issue is that nxsig_timeout can occur before we get into up_block_task. I'm not sure how to protect for this yet.
@patacongo @davids5 @xiaoxiang781216 Any ideas?

Reply:
> The issue is that nxsig_timeout can occur before we get into up_block_task. I'm not sure how to protect for this yet.

As I argued above, try using up_udelay(6).  There is no loss in performance in this case because the CPU is never relinquished in time by the thread anyway.

Reply:
> The issue is that nxsig_timeout can occur before we get into up_block_task. I'm not sure how to protect for this yet.
> @patacongo @davids5 @xiaoxiang781216 Any ideas?

nxsig_timedwait is protected by the critical section, so before up_block_task get called, the interrupt service doesn't have the chance to run.

> As I argued above, try using up_udelay(6). There is no loss in performance in this case because the CPU is never relinquished in time by the thread anyway. This corrnercase is certainly not worth a redesign.

No, if @antmerlino's report is correct, we need fix this bug:
1.Either fix the race condition or
2.Enlarge timeout to a safe value
How do you check every place to ensure all timeout isn't too small to hit the race condition?

Reply:
@patacongo Can you explain something to me.

In sig_timedwait.c line 258 we enter a critical section, disabling interrupts.

We do not call leave_critical_section(), re-enabling interrupts before we start the watchdog and call up_block_task.

How does this work? Does IRQ context switch somewhere when the task switches to re-enable interrupts?

Reply:
> 
> 
> @patacongo Can you explain something to me.
> 
> In sig_timedwait.c line 258 we enter a critical section, disabling interrupts.
> 
> We do not call leave_critical_section(), re-enabling interrupts before we start the watchdog and call up_block_task.
> 
> How does this work? Does IRQ context switch somewhere when the task switches to re-enable interrupts?

The critical section (and interrupt state) is a per thread property.  Disabling interrupts only has effect while the task is running.  When it suspends and another task runs, then that task's interrupt state is placed into effect.  So (usually) interrupts are re-enabled while the task sleeps.

I works in the context switch because the context switch saves and restores the current state of the ARM interrupt control registers.

Reply:
Okay, my report was wrong. There was never a race condition at the sig_timedwait level - Greg's answer cleared that up for me. I just didn't protect for the race condition completely at the STM32 level, there was still a very specific timing sequence that would cause it to lock up.

I've now been running stable with the latest branch.

I can't get the nxstyle line 800 error to go away. Otherwise I fixed all other formatting issues.

This is ready for final review and then merging.